segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java

```diff
@@ -577,41 +586,67 @@ private CacheIndexEntry addSingleEntryToCacheAndIndex(BufferView data, long offs
-     * more up-to-date one existing already, the given entry will be returned as rejected (and no modifications will
-     * be made to the index). A null value will be returned if there was no conflict.
+     * @param data          A {@link BufferView} representing the data to insert.
+     * @param segmentOffset The segment offset that maps to the first byte in the given {@link BufferView}.
+     * @return A {@link CacheIndexEntry} representing the index entry added. If the given {@link BufferView} spanned
+     * multiple entries (due to index fragmentation), only the last {@link CacheIndexEntry} is added.
      */
+    private CacheIndexEntry insertEntriesToCacheAndIndex(BufferView data, long segmentOffset) {
+        CacheIndexEntry lastInsertedEntry = null;
+        synchronized (this.lock) {
+            while (data != null && data.getLength() > 0) {
+                // Figure out if the first byte in the buffer is already cached.
+                ReadIndexEntry existingEntry = this.indexEntries.getFloor(segmentOffset);
+                long overlapLength;
+                if (existingEntry != null && existingEntry.getLastStreamSegmentOffset() >= segmentOffset) {
+                    // First offset exists already. We need to skip over to the end of this entry.
+                    overlapLength = existingEntry.getStreamSegmentOffset() + existingEntry.getLength() - segmentOffset;
+                    segmentOffset += overlapLength;
+                } else {
+                    // First offset does not exist. Let's find out how much we can insert.
+                    existingEntry = this.indexEntries.getCeiling(segmentOffset);
+                    overlapLength = existingEntry == null ? data.getLength() : existingEntry.getStreamSegmentOffset() - segmentOffset;
+                    assert overlapLength > 0 : "indexEntries.getFloor(offset) == null != indexEntries.getCeiling(offset)";
+
+                    // Slice the data that we need to insert. We may be able to insert the whole buffer at once.
+                    BufferView dataToInsert = overlapLength >= data.getLength() ? data : data.slice(0, (int) overlapLength);
+                    CacheIndexEntry newEntry;
+                    int dataAddress = this.cacheStorage.insert(dataToInsert);
+                    try {
+                        newEntry = new CacheIndexEntry(segmentOffset, dataToInsert.getLength(), dataAddress);
+                        ReadIndexEntry overriddenEntry = addToIndex(newEntry);
+                        assert overriddenEntry == null : "Insert overrode existing entry; " + segmentOffset + ":" + dataToInsert.getLength();
+                        lastInsertedEntry = newEntry;
+                    } catch (Throwable ex) {
+                        // Clean up the data we inserted if we were unable to add it to the index.
+                        this.cacheStorage.delete(dataAddress);
```

Comment:
Is this line reachable? (in practice)

Reply:
No. Not really. If we end up here then some exception got thrown from `addToIndex` and we must make sure we delete the data from the cache otherwise it will stay there forever. The only exception thrown from there is an `ObjectClosedException`, which indicates the whole index is about to close (or has just been) so we must clean up anyway.